File: Ellen Jung/2023-05-14-balanced-binary-tree.py
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution(object):
    def isBalanced(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        if root is None:
            return True

        left = self.get_height(root.left)
        right = self.get_height(root.right)

        if abs(left - right) > 1:
            return False

        left_balanced = self.isBalanced(root.left)
        right_balanced = self.isBalanced(root.right)

        if left_balanced and right_balanced:
            return True
        else:
            return False

    def get_height(self, root):
        if root is None:
            return 0

        left_length = self.get_height(root.left)
        right_length = self.get_height(root.right)

        if left_length > right_length:
            return left_length + 1
        else:
            return right_length + 1

File: kiwoon-jung/2023-05-14-Balanced-Binary-Tree.js
var isBalanced = function (root) {
  let flag = true;

  const getHeights = (node, height) => {
    if (!node) {
      return 0;
    }

    const leftHeight = getHeights(node.left, height + 1);
    const rightHeight = getHeights(node.right, height + 1);

    if (Math.abs(rightHeight - leftHeight) > 1) {
      flag = false;
    }

    return Math.max(leftHeight, rightHeight) + 1;
  };

  getHeights(root, 0);

  return flag;
};
